### PR TITLE
find recurring donation by the donation name in SFDC

### PR DESCRIPTION
**Pull request by LandonBaer721**

### CODE REVIEW

File: src/main/java/com/impactupgrade/nucleus/service/segment/SfdcCrmService.java [hunk 1]
```diff
@@ -66,8 +66,17 @@ public SfdcCrmService(Environment env) {
     return toCrmRecurringDonation(sfdcClient.getRecurringDonationBySubscriptionId(paymentGatewayEvent.getSubscriptionId()));
   }
 
+  @Override
   public Optional<CrmRecurringDonation> getRecurringDonation(ManageDonationEvent manageDonationEvent) throws Exception {
-    return toCrmRecurringDonation(sfdcClient.getRecurringDonationById(manageDonationEvent.getDonationId()));
+    if (!Strings.isNullOrEmpty(manageDonationEvent.getDonationId())) {
+      log.info("attempting to retrieve contact by ID {}...", manageDonationEvent.getDonationId());
```
Comment (brmeyer) [inline src/main/java/com/impactupgrade/nucleus/service/segment/SfdcCrmService.java:72]: Picky: `retrieve recurring donation`

File: src/main/java/com/impactupgrade/nucleus/service/segment/SfdcCrmService.java [hunk 2]
```diff
@@ -66,8 +66,17 @@ public SfdcCrmService(Environment env) {
...
+      return toCrmRecurringDonation(sfdcClient.getRecurringDonationById(manageDonationEvent.getDonationId()));
+    } else if (!Strings.isNullOrEmpty(manageDonationEvent.getDonationName())) {
+      log.info("attempting to retrieve contact by name {}...", manageDonationEvent.getDonationName());
```
Comment (brmeyer) [inline src/main/java/com/impactupgrade/nucleus/service/segment/SfdcCrmService.java:75]: Picky: `retrieve recurring donation`

File: src/main/java/com/impactupgrade/nucleus/model/ManageDonationEvent.java [hunk 3]
```diff
@@ -31,13 +32,17 @@ public ManageDonationEvent(Environment.RequestEnvironment requestEnv) {
...
+    if (formData.recurringDonationId.isPresent() && !Strings.isNullOrEmpty(formData.recurringDonationId.get())) this.donationId = formData.recurringDonationId.get();
+    if (formData.recurringDonationName.isPresent() && !Strings.isNullOrEmpty(formData.recurringDonationName.get())) this.donationName = formData.recurringDonationName.get();
+
     formData.stripeToken.ifPresent(s -> this.stripeToken = s);
 
-    if (!Strings.isNullOrEmpty(formData.amount.get())) this.amount = Double.parseDouble(formData.amount.get());
-    if (formData.pauseDonationUntilDate.isPresent()) this.pauseDonationUntilDate = Utils.getCalendarFromDateString(formData.pauseDonationUntilDate.get());
-    if (formData.resumeDonationOnDate.isPresent()) this.resumeDonationOnDate = Utils.getCalendarFromDateString(formData.resumeDonationOnDate.get());
-    if (!Strings.isNullOrEmpty(formData.nextPaymentDate.get())) this.nextPaymentDate = Utils.getCalendarFromDateString(formData.nextPaymentDate.get());
+    if (formData.amount != null && formData.amount.isPresent()) this.amount = formData.amount.get();
```
Comment (brmeyer) [inline src/main/java/com/impactupgrade/nucleus/model/ManageDonationEvent.java:41]: Could these ever actually be null?

**Review by brmeyer — CHANGES_REQUESTED** [review]

File: src/main/java/com/impactupgrade/nucleus/model/ManageDonationEvent.java (hunk 3 above)
Comment (LandonBaer721) [reply src/main/java/com/impactupgrade/nucleus/model/ManageDonationEvent.java:42]: IIRC when amount is not provided by the user, it arrives to this point as a null/empty value. I had expected it to arrive as an Optional.empty(). Will have to test some more to see if this is still the case.

**Pull request merged**